src/lib/UtapiReplay.js

```diff
@@ -0,0 +1,173 @@
+        if (config && config.localcache) {
+            this.localcache = new Datastore()
+                .setClient(redisClient(config.localcache, this.log));
+        } else {
+            this.localcache = new Datastore()
+                .setClient(redisClient({
+                    host: '127.0.0.1',
+                    port: 6379,
+                }, this.log));
+        }
+        if (config) {
+            this.utapiClient = new UtapiClient(config);
+        }
+        if (config && config.batchSize) {
+            this.BATCH_SIZE = config.batchSize;
+        }
+        if (config && config.replaySchedule) {
+            this.REPLAY_SCHEDULE = config.replaySchedule;
+        }
+    }
+
+    /**
+     * Sets the replay lock.
+     * @param {boolean} isLocked - The value to set `this.REPLAY_LOCK`.
+     * @param {callback} cb - Callback to call.
+     * @return {UtapiReplay} this - UtapiReplay instance.
+     */
+    _setReplayLock(isLocked) {
+        this.REPLAY_LOCK = isLocked;
+        return this;
+    }
+
+    /**
+     * Validates that all required items can be retrived from list element.
+     * @param {string} element - List element from local cache.
+     * @return {boolean} `true` if valid, `false` otherwise.
+     */
+    _validateElement(element) {
+        let data;
+        try {
+            data = JSON.parse(element);
+        } catch (err) {
+            return false;
+        }
+        if (!data.action || !data.reqUid || !data.params) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Pushes list elements using the array returned by nodeio.
+     * @param {array[]} metricsArr - The array returned by nodeio containing the
+     * an array of an error code and value for each element in the cache list.
+     * @param {callback} cb - Callback to call.
+     * @return {function} async.each - Iterates through all array elements.
+     */
+    _pushCachedMetrics(metricsArr, cb) {
+        return async.each(metricsArr, (arr, next) => {
+            const actionErr = arr[0];
+            const element = arr[1];
+            // If there is an error in one of the RPOP commands, it remains in
+            // the local cache list, so do not handle that element.
+            if (!actionErr && element) {
+                if (!this._validateElement(element)) {
+                    this.log.fatal('cannot get properties from element',
+                        { method: 'UtapiReplay._validateElement' });
+                    return next();
+                }
+                this.log.trace('pushing metric with utapiClient::pushMetric',
+                    { method: 'UtapiReplay._pushCachedMetrics' });
+                const { action, reqUid, params } = JSON.parse(element);
+                // We do not pass the callback to pushMetric since UtapiClient
+                // will handle pushing it to local cache if internal error.
+                this.utapiClient.pushMetric(action, reqUid, params);
```

Comment:
Good catch. To handle such a case, I made a change so that utapiClient uses a local Redis datastore if no configuration is given.